### PR TITLE
PLAT-908-1 Implemented New Role Concept

### PR DESCRIPTION
**Pull request by oliver-richers**
image
image
image

- planned for next PRs:
- tab for roles in EMF user popup
- tabs for users and permissions in EMF role popup
- global invalidation of permission caches

### CODE REVIEW

File: platform-core-module/src/test/java/com/softicar/platform/core/module/permission/UserModulePermissionCacheTest.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	private final UserModulePermissionCache cacheBob;
+	private final IModuleInstance<?> moduleInstance;
+
+	public UserModulePermissionCacheTest() {
+
+		this.alice = insertUser("alice");
+		this.cacheAlice = new UserModulePermissionCache(alice);
+		this.bob = insertUser("bob");
+		this.cacheBob = new UserModulePermissionCache(bob);
+		this.moduleInstance = AGCoreModuleInstance.getInstance();
```
Comment (alexander-schmidt) [inline platform-core-module/src/test/java/com/softicar/platform/core/module/permission/UserModulePermissionCacheTest.java:35]: I think we need at least one test case that involves a different/second module instance, to ensure that the module instance in fact matters to the implementation of `hasModulePermission`.

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionMap.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import com.softicar.platform.core.module.user.AGUser;
+import com.softicar.platform.core.module.uuid.AGUuid;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.UUID;
+
+public class UserModulePermissionMap {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionMap.java:17]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionCache.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import com.softicar.platform.core.module.role.user.AGRoleUser;
+import com.softicar.platform.core.module.user.AGUser;
+import java.util.UUID;
+
+public class UserModulePermissionCache {
+
+	private final DerivedObject<UserModulePermissionMap> permissions;
+
+	public UserModulePermissionCache(AGUser user) {
+
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionCache.java:17]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionMap.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	private void loadDirectAssignments() {
+
+		AGModuleInstancePermissionAssignment.TABLE//
+			.createSelect()
+			.where(AGModuleInstancePermissionAssignment.USER.isEqual(user))
+			.where(AGModuleInstancePermissionAssignment.ACTIVE.isTrue())
+			.forEach(this::addPermission);
+	}
+
+	private void loadRolePermissions() {
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionMap.java:47]: did so, but I am not sure if the negligible performance increase outweighs the increased complexity

File: platform-core-module/src/main/java/com/softicar/platform/core/module/role/RoleManagementPage.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	@Override
+	protected IEmfTable<?, ?, AGCoreModuleInstance> getTable() {
+
+		return AGRole.TABLE;
+	}
+
+	@Override
+	public IEmfModulePermission<AGCoreModuleInstance> getRequiredPermission() {
+
+		return CorePermissions.SUPER_USER;
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/role/RoleManagementPage.java:37]: true, changed to ACCESS_MANAGEMENT, also for the other 2 pages

File: platform-core-module/src/main/java/com/softicar/platform/core/module/role/permission/RolePermissionManagementPage.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	@Override
+	protected IEmfTable<?, ?, AGCoreModuleInstance> getTable() {
+
+		return AGRolePermission.TABLE;
+	}
+
+	@Override
+	public IEmfModulePermission<AGCoreModuleInstance> getRequiredPermission() {
+
+		return CorePermissions.SUPER_USER;
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/role/permission/RolePermissionManagementPage.java:37]: see above

File: platform-core-module/src/main/java/com/softicar/platform/core/module/role/user/RoleUserManagementPage.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	@Override
+	protected IEmfTable<?, ?, AGCoreModuleInstance> getTable() {
+
+		return AGRoleUser.TABLE;
+	}
+
+	@Override
+	public IEmfModulePermission<AGCoreModuleInstance> getRequiredPermission() {
+
+		return CorePermissions.SUPER_USER;
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/role/user/RoleUserManagementPage.java:37]: see above

File: platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionMap.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+	private final AGUser user;
+	private final Map<AGModuleInstance, Set<UUID>> permissions;
+
+	public UserModulePermissionMap(AGUser user) {
+
+		this.user = user;
+		this.permissions = new TreeMap<>();
+
+		loadDirectAssignments();
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/permission/UserModulePermissionMap.java:27]: Sorry, but I see no advantage. AFAIK the c'tor is meant to establish a usable state of the object. We are not doing any semantic side effects, so what is the point?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/role/permission/AGRolePermission.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.softicar.platform.common.core.i18n.IDisplayString;
+import com.softicar.platform.emf.object.IEmfObject;
+
+public class AGRolePermission extends AGRolePermissionGenerated implements IEmfObject<AGRolePermission> {
+
+	@Override
+	public IDisplayString toDisplayWithoutId() {
+
+		return IDisplayString.format("%s - %s", getRole().toDisplayWithoutId(), getPermissionUuid().toDisplayWithoutId());
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/role/permission/AGRolePermission.java:11]: changed, but I still prefer the dash (but does not really matter, this method will hardly ever be called)

File: platform-core-module/src/main/java/com/softicar/platform/core/module/role/user/AGRoleUser.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.softicar.platform.common.core.i18n.IDisplayString;
+import com.softicar.platform.emf.object.IEmfObject;
+
+public class AGRoleUser extends AGRoleUserGenerated implements IEmfObject<AGRoleUser> {
+
+	@Override
+	public IDisplayString toDisplayWithoutId() {
+
+		return IDisplayString.format("%s - %s", getRole().toDisplayWithoutId(), getUser().toDisplayWithoutId());
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/role/user/AGRoleUser.java:11]: changed, but I still prefer the dash (but does not really matter, this method will hardly ever be called)

**Review by alexander-schmidt — DISMISSED** [review]

**Review by alexander-schmidt — APPROVED** [review]

**Pull request merged**